File: PrefixSum/minPenaltyForShop.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    int bestClosingTime(string customers) {
        int n = customers.size();
        vector<int>preSum(n+1,0);
        vector<int>sufSum(n+1,0);
        vector<int>overallPenalty(n+1,0);
        for(int i=1;i<=n;i++){
            preSum[i] += preSum[i-1];
            if(customers[i-1]=='N')preSum[i] += 1;
        }
        for(int i=n-1;i>=0;i--){
            sufSum[i] += sufSum[i+1];
            if(customers[i]=='Y')sufSum[i] += 1;
        }
        for(int i=0;i<=n;i++){
            overallPenalty[i] = preSum[i]+sufSum[i];
        }
        int minIdx = 0;
        int idx = 0;
        int minPenalty = overallPenalty[0];
        while(idx<=n){
            if(minPenalty > overallPenalty[idx]){
                minIdx = idx;
                minPenalty = overallPenalty[idx];
            }
            idx += 1;
        }
        return minIdx;
    }
};

File: PrefixSum/minPenaltyForShop.py
class Solution:
    def bestClosingTime(self, customers: str) -> int:
        n = len(customers)
        sufSum = [0]*(n+1)
        preSum = [0]*(n+1)
        idx = n-1
        while idx >= 0:
            sufSum[idx] += sufSum[idx+1]
            if customers[idx]=='Y':
                sufSum[idx] += 1
            idx -= 1
        idx = 1
        while idx <= n:
            preSum[idx] += preSum[idx-1]
            if customers[idx-1]=='N':
                preSum[idx] += 1
            idx += 1
        overallPenalty = [0] * (n+1)
        for i in range(n+1):
            overallPenalty[i] = preSum[i] + sufSum[i]
        minIdx = 0
        idx = 0
        minPenalty = overallPenalty[0]
        while idx <= n:
            if minPenalty > overallPenalty[idx]:
                minIdx = idx
                minPenalty = overallPenalty[idx]
            idx += 1
        return minIdx